db/fetcher/indexer_iterators.go

```diff
@@ -520,9 +523,98 @@ func getFieldsBytes(conditions []fieldFilterCond) ([][]byte, error) {
 	return result, nil
 }
 
+func (f *IndexFetcher) newPrefixIndexIterator(
+	fieldConditions []fieldFilterCond,
+	matchers []valueMatcher,
+) (*eqPrefixIndexIterator, error) {
+	keyFieldValues := make([][]byte, 0, len(fieldConditions))
+	for i := range fieldConditions {
+		if fieldConditions[i].op != opEq {
+			break
+		}
+
+		fieldVal := client.NewFieldValue(client.LWW_REGISTER, fieldConditions[i].val)
+
+		keyValueBytes, err := fieldVal.Bytes()
+		if err != nil {
+			return nil, err
+		}
+
+		keyFieldValues = append(keyFieldValues, keyValueBytes)
+	}
+
+	// iterators for _eq filter already iterate over keys with first field value
+	// matching the filter value, so we can skip the first matcher
+	if len(matchers) > 1 {
+		matchers[0] = &anyMatcher{}
+	}
+
+	indexKey := f.newIndexDataStoreKey()
+	indexKey.FieldValues = keyFieldValues
+	return &eqPrefixIndexIterator{
+		indexKey: indexKey,
+		execInfo: &f.execInfo,
+		matchers: matchers,
+	}, nil
+}
+
+func (f *IndexFetcher) newInIndexIterator(
+	fieldConditions []fieldFilterCond,
+	matchers []valueMatcher,
+) (*inIndexIterator, error) {
+	inArr, ok := fieldConditions[0].val.([]any)
+	if !ok {
+		return nil, ErrInvalidInOperatorValue
+	}
+	keyFieldArr := make([][]byte, 0, len(inArr))
+	for _, v := range inArr {
+		fieldVal := client.NewFieldValue(client.LWW_REGISTER, v)
+		keyFieldBytes, err := fieldVal.Bytes()
+		if err != nil {
+			return nil, err
+		}
+		keyFieldArr = append(keyFieldArr, keyFieldBytes)
+	}
+
+	// iterators for _in filter already iterate over keys with first field value
+	// matching the filter value, so we can skip the first matcher
+	if len(matchers) > 1 {
+		matchers[0] = &anyMatcher{}
```

Comment:
suggestion: `matchers` is not owned by this function, seeing an element in it overwritten like this sets off alarm bells in my head.  Especially when this function's signature does not suggest that it will be mutating any of its input params.

If you wish to keep the mutation, can I please ask that you document it on the function signature?

The same goes for `newPrefixIndexIterator` which does similar.

Reply:
added a comment